@lingui/macro has undeclared dependency on @lingui/react

**Describe the bug**

You can't use `@lingui/macro` without installing react, because packages/macro/index.d.ts pulls in `react` and `@lingua/react`:

https://github.com/lingui/js-lingui/blob/4ad90d4c60b752e9f0d82f3dd71936711c51573d/packages/macro/index.d.ts#L1-L3

However, the dependencies are not listed in package.json.

**To Reproduce**

Install `@lingui/core` and `@lingui/macro`, _but no other packages_ into a (pure) TypeScript project and try to use:

```typescript
import { t } from "@lingui/macro"
```

**Expected behavior**

Either the dependency should be made clear or the package should be usable without react.

**Additional context**
Add any other context about the problem here.

- jsLingui version ``lingui --version``: 3.3.0

I use:

- typescript 4.1.3
- browserify 17.0.0
- tsify 5.0.2
- babelify 10.0.0
- babel 7.12.10

with the browserify / tsify / babelify pipeline converting from typescript to browser-ready code. `tsconfig.js` contains:

```json
{
  "compilerOptions": {
    "target": "esnext",
    "moduleResolution": "node",
    "module": "es6",
    "allowJs": true,
    "strict": true,
    "isolatedModules": true,
    "esModuleInterop": true
  }
}
```

(plus project-specific paths for `types`, `rootDir` and an `exclude` list next to the `compilerOptions`), and `babel.config.json` contains:

```json
{
  "presets": ["@babel/env", "@babel/typescript"],
  "plugins": [
    "@babel/proposal-class-properties",
    "@babel/proposal-object-rest-spread",
    ["polyfill-regenerator", { "method": "usage-pure" }]
  ]
}
```

If I use:

```typescript
import { t } from '@lingui/macro'

const SOME_CODE = {
  hi: t`example`
}
```

and add `macros` to my `babel.conf.json` config `plugins` list, I can't compile anymore, I get an error:

```none
TypeScript error: /.../node_modules/@lingui/macro/index.d.ts(1,61): Error TS2307: Cannot find module 'react' or its corresponding type declarations.
```

So I do have to install react, I can't compile without it being installed.

I do understand that using `import type ...` only depends on the type information at compile time, no runtime dependency is pulled in. I can install `react` and `@lingui/react` and not have it end up in my output, but then you do need to list `react` as a dependency! I don't want to have to find out that I have to install react manually here.

I didn't have time until now to investigate further. The difference between your test and my setup is a setting that `tsc --init` has set in `tsconfig.json`:

```json
{
  "compilerOptions": {
    // ...
    /* Advanced Options */
    "skipLibCheck": true,                           /* Skip type checking of declaration files. */
    // ...
}
```

See the [`skipLibCheck` docs](https://www.typescriptlang.org/tsconfig#skipLibCheck):

> Skip type checking of declaration files.
>
> This can save time during compilation at the expense of type-system accuracy. *[...]* Rather than doing a full check of all `d.ts` files, TypeScript will type check the code you specifically refer to in your app’s source code.

If you set that to the default `false`, you'll see the same errors:

```sh
$ tsc  # no errors
$ sed -i '' -e 's/"skipLibCheck": true, /"skipLibCheck": false,/' tsconfig.json
$ tsc
node_modules/@lingui/macro/index.d.ts:1:61 - error TS2307: Cannot find module 'react' or its corresponding type declarations.

1 import type { ReactElement, ComponentType, ReactNode } from "react"
                                                              ~~~~~~~

node_modules/@lingui/macro/index.d.ts:2:40 - error TS2307: Cannot find module '@lingui/core' or its corresponding type declarations.

2 import type { MessageDescriptor } from "@lingui/core"
                                         ~~~~~~~~~~~~~~

node_modules/@lingui/macro/index.d.ts:3:39 - error TS2307: Cannot find module '@lingui/react' or its corresponding type declarations.

3 import type { TransRenderProps } from "@lingui/react"
                                        ~~~~~~~~~~~~~~~


Found 3 errors.
```

--- a/docs/guides/typescript.rst
+++ b/docs/guides/typescript.rst
@@ -2,7 +2,7 @@
 Typescript
 **********
 
-Lingui supports typescript out of the box since version ``3.0.0``. Feel free to submit any query you find related to typescript on Github Issues
+Lingui supports typescript types out of the box since version ``3.0.0``. Feel free to submit any query you find related to typescript on Github Issues
 
 Webpack setup
 =============
@@ -61,6 +61,29 @@ otherwise compiled catalogs can't be imported using ES ``import``, but rather Co
   {
     "compileNamespace": "ts"
   }
+
+
+Macros types in non-React environments
+======================================
+
+Since the opening of this issue we investigated that macros can be used on Typescript environments where React isn't required.
+
+Now we're shipping two declaration types:
+  - ``index.d.ts`` files with ``@lingui/core``, ``@lingui/react`` and ``react`` as peerDependencies.
+  - ``global.d.ts`` files with just ``@lingui/core`` as peerDependencies.
+
+Now you can modify your ``tsconfig.json`` in your root directory and reference the global file:
+
+.. code-block:: json
+
+  {
+    "compilerOptions": {
+      "types": [
+        "./node_modules/@lingui/macro/global",
+      ]
+    }
+  }
+
 
 Type definitions
 ================

--- a/packages/macro/global.d.ts
+++ b/packages/macro/global.d.ts
@@ -1,0 +1,63 @@
+declare module '@lingui/macro' {
+  import type { MessageDescriptor } from "@lingui/core"
+
+  export type BasicType = {
+    id?: string
+    comment?: string
+  }
+
+  export function t(
+    literals: TemplateStringsArray | MessageDescriptor,
+    ...placeholders: any[]
+  ): string
+
+  export type UnderscoreDigit<T = string> = { [digit: string]: T }
+  export type ChoiceOptions<T = string> = {
+    offset?: number
+    zero?: T
+    one?: T
+    few?: T
+    many?: T
+    other?: T
+  } & UnderscoreDigit<T>
+
+  export function plural(arg: number | string, options: ChoiceOptions & BasicType): string
+  export function selectOrdinal(
+    arg: number | string,
+    options: ChoiceOptions & BasicType
+  ): string
+  export function select(arg: string, choices: Record<string, string> & BasicType): string
+  export function defineMessages<M extends Record<string, MessageDescriptor>>(
+    messages: M
+  ): M
+  export function defineMessage(descriptor: MessageDescriptor): MessageDescriptor
+
+  export type ChoiceProps = {
+    value?: string | number
+  }  & ChoiceOptions<string>
+
+  /**
+   * The types should be changed after this PR is merged
+   * https://github.com/Microsoft/TypeScript/pull/26797
+   *
+   * then we should be able to specify that key of values is same type as value.
+   * We would be able to remove separate type Values = {...} definition
+   * eg.
+   * type SelectProps<Values> = {
+   *  value?: Values
+   *  [key: Values]: string
+   * }
+   *
+   */
+  type Values = { [key: string]: string }
+
+  export type SelectProps = {
+    value: string
+    other: any
+  } & Values
+
+  export const Trans: any
+  export const Plural: any
+  export const Select: any
+  export const SelectOrdinal: any
+}

--- a/packages/macro/package.json
+++ b/packages/macro/package.json
@@ -25,6 +25,7 @@
     "LICENSE",
     "README.md",
     "*.js",
+    "global.d.ts",
     "index.d.ts"
   ],
   "dependencies": {

--- a/scripts/build/babel.js
+++ b/scripts/build/babel.js
@@ -30,6 +30,10 @@ function walk(base, relativePath = "") {
   return files
 }
 
+function declarationFinder(typesPath) {
+  return typesPath.map(p => fs.existsSync(p) && path.basename(p)).filter(Boolean)
+}
+
 module.exports = async function(bundle) {
   const logKey = chalk.white.bold(bundle.entry)
 
@@ -39,14 +43,13 @@ module.exports = async function(bundle) {
     const resolvedEntry = require.resolve(bundle.entry)
     const packageDir = path.dirname(resolvedEntry)
     const srcDir = path.join(packageDir, "src")
-
-    const declarationFilePath = path.join(packageDir, "index.d.ts")
-
-    const files = walk(srcDir)
-
-    if (fs.existsSync(declarationFilePath)) {
-      files.push("index.d.ts")
-    }
+    const files = [
+      ...walk(srcDir),
+      ...declarationFinder([
+        path.join(packageDir, "index.d.ts"),
+        path.join(packageDir, "global.d.ts"),
+      ])
+    ]
 
     for (const filename of files) {
       const [mainOutputPath] = Packaging.getBundleOutputPaths(
@@ -65,7 +68,7 @@ module.exports = async function(bundle) {
         )
         fs.writeFileSync(mainOutputPath.replace(/\.ts$/, ".js"), code)
       } else {
-        fs.copyFileSync(path.join(packageDir, filename), mainOutputPath)
+        fs.copyFileSync(path.join(packageDir, filename), path.join(outputDir, filename))
       }
     }
   } catch (error) {

